Cssp smartcard_creds

informatimago opened the pull request:
Implemented:
- obtaining X509 user identities from IAS-ECC smartcards (scquery),
- obtaining a Ticket Granting Ticket from Kerberbos (kinit) based on the above X509 user identity,
- sending to the server TSCredentials of type TSSmartcardCreds.
- The TSRemoteGuardCreds credentials and ASN.1 encoding is also implemented, but not used yet (other protocols (eg. RDPEAR) need to be implemented to use Remote Guard).

Example of invocation:
```
 ./client/X11/xfreerdp /v:${host} /d:${domain} /smartcard /smartcard-logon /pkcs11-module:/usr/local/lib/libiaspkcs11.so /pkinit-anchors:/etc/chain1.pem,/etc/chain2.pem /csp:'Middleware IAS ECC Cryptographic Provider'
```

akallabeth commented on channels/smartcard/client/smartcard_operations.c:
```diff
@@ -660,16 +660,40 @@ static LONG smartcard_GetStatusChangeA_Decode(SMARTCARD_DEVICE* smartcard,
 }
 
+
+#define CARD_GET_STATUS_CHANGE(call, operation, ret, status, status_change)			\
```
Better use a static inline function for this.

informatimago replied:
Indeed, there are several modules, but they make sense only when used in combination.  
Is it ok for you if I split the PR module by module, even if they remain dead-code until everything is merged?

bmiklautz commented:
@freerdp-bot test

akallabeth commented on winpr/libwinpr/utils/wlog/wlog.c:
```diff
@@ -346,7 +346,8 @@ static BOOL WLog_WritePacket(wLog* log, wLogMessage* message)
 
 BOOL WLog_PrintMessageVA(wLog* log, DWORD type, DWORD level, DWORD line,
-                         const char* file, const char* function, va_list args)
+                         const char* file, const char* function,
```
Why this change?

informatimago replied:
To be able to call this function from the scquery.c report_level function that has a format string and a va_list arguments.
The format argument to WLog_PrintMessageVA is mandatory, while we could pass no argument for the va_list (it can be empty). Therefore it is safer to explicitely define it.

akallabeth commented on channels/smartcard/client/smartcard_pack.c:
```diff
@@ -309,21 +312,26 @@ LONG smartcard_unpack_redir_scard_context_ref(SMARTCARD_DEVICE* smartcard, wStre
 LONG smartcard_pack_redir_scard_context_ref(SMARTCARD_DEVICE* smartcard, wStream* s,
         REDIR_SCARDCONTEXT* context)
 {
-	Stream_Write_UINT32(s, context->cbContext); /* Length (4 bytes) */
```
Are you sure you can remove the lenght field if there is no data?

akallabeth commented on channels/smartcard/client/smartcard_pack.c:
```diff
@@ -309,21 +312,26 @@ LONG smartcard_unpack_redir_scard_context_ref(SMARTCARD_DEVICE* smartcard, wStre
-	if (context->cbContext)
+	if (context->cbContext > 0)
 	{
+		Stream_Write_UINT32(s, context->cbContext); /* Length (4 bytes) */
```
Are you sure you can remove the lenght field if there is no data?

akallabeth commented on channels/smartcard/client/smartcard_pack.c:
```diff
@@ -409,10 +417,11 @@ LONG smartcard_unpack_redir_scard_handle_ref(SMARTCARD_DEVICE* smartcard, wStrea
 LONG smartcard_pack_redir_scard_handle_ref(SMARTCARD_DEVICE* smartcard, wStream* s,
         REDIR_SCARDHANDLE* handle)
 {
-	Stream_Write_UINT32(s, handle->cbHandle); /* Length (4 bytes) */
```
Are you sure you can remove the lenght field if there is no data?

informatimago replied:
> The image you committed is corrupted and can not be read.

I can visualize this docs/Smartcard-Logon-Dataflow.png on firefox or safari from github.
But I will make it a graphviz dot file, so it’ll be textual and easier to maintain.

akallabeth commented on client/common/cmdline.h:
```diff
@@ -166,6 +166,15 @@ static COMMAND_LINE_ARGUMENT_A args[] =
 	{ "smart-sizing", COMMAND_LINE_VALUE_OPTIONAL, "<width>x<height>", NULL, NULL, -1, NULL, "Scale remote desktop to window size" },
 	{ "smartcard", COMMAND_LINE_VALUE_OPTIONAL, "<str>[,<str>...]", NULL, NULL, -1, NULL, "Redirect the smartcard devices containing any of the <str> in their names." },
 	{ "smartcard-logon", COMMAND_LINE_VALUE_FLAG, NULL, NULL, NULL, -1, NULL, "Activates Smartcard Logon authentication. (EXPERIMENTAL: NLA not supported)" },
+        { "pin", COMMAND_LINE_VALUE_OPTIONAL, "<PIN code>", NULL, NULL, -1, NULL, "PIN code" },
```
Maybe add these as arguments to `smartcard-logon` instead of having so many new arguments?

informatimago commented:
> * Would suggest splitting it in smaller ones each targetting a specific change
>   (winpr, nla, smartcard, ...)

Indeed, there are several modules, but they make sense only when used in combination.
Is it ok for you if I split the PR module by module, even if they remain dead-code until everything is merged?

akallabeth commented:
Any specific reason you are using `dlopen` to access the `pkcs11` module?

akallabeth commented:
Looking at the code in `scquery` it looks like a bunch of stuff mixed from various sources, with redefinitions of really a lot of stuff. Definitely needs cleanup.
Same for kerberos, why copy headers and stuff in there? A proper CMake `find_library` or similar should suffice.

akallabeth commented:
The image you committed is corrupted and can not be read.

informatimago commented:
> Any specific reason you are using `dlopen` to access the `pkcs11` module?

There are several pkcs11 modules, and we want to let the user choose which one to use. For example, the current customer uses the closed source  libiaspkcs11 module, with IAS-ECC smartcards.
https://ants.gouv.fr/Les-solutions/Identite-numerique2/Technologie-cartes-IAS-ECC/Pilote-carte-middleware

But with other smartcards we would use opencs-pkcs11.

This depends on the smartcard and on the service we connect to.

informatimago commented:
> Looking at the code in `scquery` it looks like a bunch of stuff mixed from various sources, with redefinitions of really a lot of stuff. Definitely needs cleanup.
> Same for kerberos, why copy headers and stuff in there? A proper CMake `find_library` or similar should suffice.

For scquery, this is possible. Or perhaps, the upstream could be delivered as a library.
For kinit, it is a command in the sources of MIT kerberos. It is not packaged as a library.
In both cases, besides removing the main, there have been some modifications, both functional (eg. extracting the CanonicalizedUser), or to integrate the sources to FreeRDP.

My original intent was actually to keep them as external tool, and to obtain the information needed from them and pass it as argument to freerdp. I integrated them because I wasn’t sure it would be possible to keep them separate, but now I have more information and it seems to me that it should be possible, and that it moght be better, mostly because of the variations in smartcard formats, and in the pkcs11 middleware modules.  To avoid integrating kinit sources in freerdp, I would have to propose a patch to MIT Kerberos kinit, so that it reports the CanonicalizedUser. Currently this could be recovered only with klist, which would be ambiguous when several tgt are stored in the cache.  Also, I think that keeping kimit separate will be better when we implement remote guard credentials for sso.

akallabeth commented:
* for the `strlist` changes, I'd suggest using `size_t` instead of `int` for sizes (and `BOOL` where appropriate)
* The changes to the command line parser look odd, would remove them
* The changes to `wlog` look odd, what was the intention?
* The changes in `channels/smartcard` look like some merge went wrong? A previously removed function is there again, some fixes (null checks, ...) have been moved/changed/... Might be a good idea to recheck them.

akallabeth commented:
Last thing I'd like to add is the command line interface.
I don't like adding so many options (we already have quite a zoo of these).
To have the same functionality without the mess, I'd propose something like `/smartcard-logon:pin:1234,start-time:monday,...`

Oh, and please remove the `T` option (along with the `Krb5Trace` setting), the same can be achieved with `/log-level:debug` or `/log-filters:your-kerberos-logger-tag:debug` (Just check if the logger you're using has level `debug` or `trace` - whichever you prefer - to enable the trace in `get_info_smartcard`)

compuguy commented:
Is this pull request related to #3946 by chance @informatimago?

akallabeth commented:
@compuguy it is an extension. The first pull allowes logon with rdp/tls authentication, this is an extension to use the NLA authentication mechanism.
Needs some rework though.

akallabeth commented:
@informatimago we're nearing 2.0 release and would like to add this to the release after that.
would it be possible to clean up this pull (and maybe split it to library changes and new feature pulls)?

martinl commented:
Is this smartcard logon dataflow diagram from informatimago:cssp-smartcard branch docs valid going forward - what are the pkinit vs kinit procedure differences:
![Smartcard-Logon-Dataflow](https://user-images.githubusercontent.com/148686/61525273-b7c00180-aa20-11e9-9614-4ad6a0e203f7.png)

akallabeth commented:
@martinl the command line options need rework, having an extra option for each parameter is not really working out.
Better would be something like `/smartcard:<option1>:<value1>,<option2>:<value2>,...` to avoid cluttering the top level options.

So no, the diagram needs some reworking.
So, what needs to be done for this:
* Fix command line, simplify it 
* Split out the dependent libraries (kerberos related)
* Maybe split up this pr in one that adds the changes in the core library (`Smartcard Credentials`, if possible with a small test case that can be run) and one adding the client side support (including the command line stuff)
* See if this `strlst` stuff is really necessary - looks like duplication of functionality already available from `C lib`, `FreeRDP` or `WinpPR`

informatimago commented:
> On 19 Jul 2019, at 11:39, martinl <notifications@github.com> wrote:
> 
> Is this smartcard logon dataflow diagram from informatimago:cssp-smartcard branch docs valid going forward - what are the pkinit vs kinit procedure differences:
>  <https://user-images.githubusercontent.com/148686/61525273-b7c00180-aa20-11e9-9614-4ad6a0e203f7.png>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/FreeRDP/FreeRDP/pull/5195?email_source=notifications&email_token=AAGBJDCMJ7OABHNGVI3SPU3QAGDURA5CNFSM4GQD4O2KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2LEPUA#issuecomment-513165264>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAGBJDB4RXGPZE25T32QLMTQAGDURANCNFSM4GQD4O2A>.
> 


pkinit is a kinit plug-in implementing Kerberos pre-authentication.

It’s activated with kinit(1) using the -X option with the X509_user_identity attribute (and X509_anchor etc).
It is to use this mode that I write scquery, to retrieve from IAS-ECC SmartCard the authentication certificates and UPN stored in those certificates, needed to build this X509_user_identity.

Notably, the way you retrieve the X509_user_identity will depend on the kind of smartcard. The sckinit script retrieves it from the IAS-ECC smartcard, but if you use other kinds of smartcard, you would have to write a different script.

https://github.com/informatimago/scquery <https://github.com/informatimago/scquery>
https://github.com/informatimago/scquery/blob/master/scripts/sckinit

Dorian implemented pkinit inside freerdp in his branch, but I don’t think it’s necessary since it’s already implemented in kinit, or a good idea for the reason mentioned above. 
We only need to retrieve the CanonicalizedUserHint from kinit, and pass it to freerdp as a command line argument.


-- 
__Pascal J. Bourguignon__

martinl commented:
Thank you for the clarification of current state of smartcard logon support.

I did some initial work on setting up my macOS Mojave test environment - rebased the informatimago:cssp-smartcard branch to current FreeRDP master and built it:

```
# clone and patch informatimago:cssp-smartcard branch
git clone https://github.com/informatimago/FreeRDP.git
cd FreeRDP
git checkout cssp-smartcard
git remote add upstream https://github.com/FreeRDP/FreeRDP.git
git fetch upstream
git rebase upstream/master

# some manual merges

# apply cmakefiles patch
patch -p1 < ../bmiklautz_pkcs11h_gssapi.diff

# install freerdp macOS dependencies
brew install opensc pkcs11-helper ffmpeg krb5

# add brew krb5 to path and compiler flags
# FIXME: build against official macOS kerberos/use pkgconfig
export PATH="/usr/local/opt/krb5/bin:$PATH"
export PATH="/usr/local/opt/krb5/sbin:$PATH"
export LDFLAGS="-L/usr/local/opt/krb5/lib"
export CPPFLAGS="-I/usr/local/opt/krb5/include"

# set pkgconfig path to brew openssl
export PKG_CONFIG_PATH=/usr/local/opt/openssl/lib/pkgconfig

# clean up cmake cached files (only needed for repeated build testing)
#find . -name CMakeCache.txt | xargs rm
#find . -name CMakeFiles | xargs rm -r

# generate makefiles
export GSS_DIR=/usr/local/opt/krb5
export GSS_ROOT_FLAVOUR=MIT
cmake -DWITH_PCSC=ON -DWITH-FFMPEG=ON -DWITH_PKCS11H=ON -DWITH_GSSAPI=ON .

# build project
cmake --build .

# get buildconfig output
./client/Mac/cli/MacFreeRDP.app/Contents/MacOS/MacFreeRDP /buildconfig 2>&1 > buildconfig.txt
```

```
cat buildconfig.txt 
This is FreeRDP version 2.0.0-dev5 (da5355e90)
Build configuration: BUILD_TESTING=OFF BUILTIN_CHANNELS=ON HAVE_AIO_H=1 HAVE_EXECINFO_H=1 HAVE_FCNTL_H=1 HAVE_FLAG_SEARCH_PATHS_FIRST=1 HAVE_INTTYPES_H=1 HAVE_MATH_C99_LONG_DOUBLE=1 HAVE_POLL_H=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIB= HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIBS= HAVE_PTHREAD_MUTEX_TIMEDLOCK_SYMBOL= HAVE_SYSLOG_H=1 HAVE_SYS_EVENTFD_H= HAVE_SYS_FILIO_H=1 HAVE_SYS_MODEM_H= HAVE_SYS_SELECT_H=1 HAVE_SYS_SOCKIO_H=1 HAVE_SYS_STRTIO_H= HAVE_SYS_TIMERFD_H= HAVE_TM_GMTOFF=1 HAVE_UNISTD_H=1 WITH_CCACHE=ON WITH_CHANNELS=ON WITH_CLIENT=ON WITH_CLIENT_AVAILABLE=1 WITH_CLIENT_CHANNELS=ON WITH_CLIENT_CHANNELS_AVAILABLE=1 WITH_CLIENT_COMMON=ON WITH_CLIENT_INTERFACE=OFF WITH_CUPS=OFF WITH_DEBUG_ALL=OFF WITH_DEBUG_CAPABILITIES=OFF WITH_DEBUG_CERTIFICATE=OFF WITH_DEBUG_CHANNELS=OFF WITH_DEBUG_CLIPRDR=OFF WITH_DEBUG_DVC=OFF WITH_DEBUG_KBD=OFF WITH_DEBUG_LICENSE=OFF WITH_DEBUG_MUTEX=OFF WITH_DEBUG_NEGO=OFF WITH_DEBUG_NLA=OFF WITH_DEBUG_NTLM=OFF WITH_DEBUG_RAIL=OFF WITH_DEBUG_RDP=OFF WITH_DEBUG_RDPDR=OFF WITH_DEBUG_RDPEI=OFF WITH_DEBUG_REDIR=OFF WITH_DEBUG_RFX=OFF WITH_DEBUG_RINGBUFFER=OFF WITH_DEBUG_SCARD=OFF WITH_DEBUG_SND=OFF WITH_DEBUG_SVC=OFF WITH_DEBUG_SYMBOLS=OFF WITH_DEBUG_THREADS=OFF WITH_DEBUG_TIMEZONE=OFF WITH_DEBUG_TRANSPORT=OFF WITH_DEBUG_TSG=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_WND=OFF WITH_DEBUG_X11=OFF WITH_DEBUG_X11_CLIPRDR=OFF WITH_DEBUG_X11_LOCAL_MOVESIZE=OFF WITH_DEBUG_XV=OFF WITH_DSP_EXPERIMENTAL=OFF WITH_DSP_FFMPEG=OFF WITH_FAAC=OFF WITH_FAAD2=OFF WITH_FFMPEG=TRUE WITH_FFMPEG=TRUE WITH_GFX_H264=ON WITH_GPROF=OFF WITH_GSM=OFF WITH_GSSAPI=ON WITH_GSTREAMER_0_10=OFF WITH_GSTREAMER_1_0=OFF WITH_ICU=OFF WITH_JPEG=OFF WITH_LAME=OFF WITH_LIBRARY_VERSIONING=ON WITH_LIBSYSTEMD=OFF WITH_MACAUDIO=ON WITH_MACAUDIO_AVAILABLE=1 WITH_MANPAGES=ON WITH_MBEDTLS=OFF WITH_OPENH264=OFF WITH_OPENSSL=ON WITH_PCSC=ON WITH_PKCS11H=ON WITH_PROFILER=OFF WITH_PULSE=OFF WITH_SAMPLE=OFF WITH_SANITIZE_ADDRESS=OFF WITH_SANITIZE_ADDRESS_AVAILABLE=1 WITH_SANITIZE_MEMORY=OFF WITH_SANITIZE_MEMORY_AVAILABLE=1 WITH_SANITIZE_THREAD=OFF WITH_SANITIZE_THREAD_AVAILABLE=1 WITH_SERVER=OFF WITH_SERVER_INTERFACE=ON WITH_SMARTCARD_INSPECT=OFF WITH_SOXR=OFF WITH_SSE2=ON WITH_THIRD_PARTY=OFF WITH_VALGRIND_MEMCHECK=OFF WITH_VALGRIND_MEMCHECK_AVAILABLE=1 WITH_X11=OFF WITH_X264=OFF WITH_ZLIB=ON
Build type:          Release
CFLAGS:               -Wno-unused-parameter -Wno-unused-macros -Wno-padded -Wno-c11-extensions -Wno-gnu -Wno-unused-command-line-argument -Wno-deprecated-declarations -DWINPR_DLL
Compiler:            Clang, 10.0.1.10010046
Target architecture: x64
```

I also built https://github.com/akallabeth/FreeRDP.git smartcard_cleanup_and_fixes branch following the same instructions

I'm planning to set up my Debian 10 build/test environment next and after that Windows Server 2016 eval for doing some actual testing with Yubikey NEO in PIV mode

martinl commented:
Debian 10 build notes:

```
# FreeRDP Debian 10 build notes
# based on https://github.com/FreeRDP/FreeRDP/wiki/Compilation

# clone master git repo
git clone https://github.com/FreeRDP/FreeRDP.git
cd FreeRDP

# install debian 10 dependencies
sudo apt-get install ninja-build build-essential git-core debhelper cdbs dpkg-dev autotools-dev cmake pkg-config xmlto libssl-dev docbook-xsl xsltproc libxkbfile-dev libx11-dev libwayland-dev libxrandr-dev libxi-dev libxrender-dev libxext-dev libxinerama-dev libxfixes-dev libxcursor-dev libxv-dev libxdamage-dev libxtst-dev libcups2-dev libpcsclite-dev libasound2-dev libpulse-dev libjpeg-dev libgsm1-dev libusb-1.0-0-dev libudev-dev libdbus-glib-1-dev uuid-dev libxml2-dev libgstreamer1.0-dev libgstreamer1.0-dev libgstreamer-plugins-base1.0-dev libgstreamer-plugins-base1.0-dev

# more build dependencies
sudo apt-get install libcunit1-dev libdirectfb-dev xmlto doxygen libxtst-dev libcairo2-dev libavcodec-dev libavutil-dev libswresample-dev

# add kerberos and pkcs11
sudo apt install libpkcs11-helper1-dev libkrb5-dev

export GSS_ROOT_FLAVOUR=MIT
cmake -DWITH_PCSC=ON -DWITH-FFMPEG=ON -DWITH_PKCS11H=ON -DWITH_GSSAPI=ON .

# build project
cmake --build .

```
get buildconfig
```
./client/X11/xfreerdp /buildconfig > buildconfig.txt

cat buildconfig.txt 
Build configuration: BUILD_TESTING=OFF BUILTIN_CHANNELS=ON HAVE_AIO_H=1 HAVE_EXECINFO_H=1 HAVE_FCNTL_H=1 HAVE_INTTYPES_H=1 HAVE_MATH_C99_LONG_DOUBLE=1 HAVE_POLL_H=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK=ON HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIB=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK_SYMBOL= HAVE_SYSLOG_H=1 HAVE_SYS_EVENTFD_H=1 HAVE_SYS_FILIO_H= HAVE_SYS_MODEM_H= HAVE_SYS_SELECT_H=1 HAVE_SYS_SOCKIO_H= HAVE_SYS_STRTIO_H= HAVE_SYS_TIMERFD_H=1 HAVE_TM_GMTOFF=1 HAVE_UNISTD_H=1 HAVE_XI_TOUCH_CLASS=1 WITH-FFMPEG=ON WITH_ALSA=ON WITH_CAIRO=OFF WITH_CCACHE=ON WITH_CHANNELS=ON WITH_CLANG_FORMAT=ON WITH_CLIENT=ON WITH_CLIENT_AVAILABLE=1 WITH_CLIENT_CHANNELS=ON WITH_CLIENT_CHANNELS_AVAILABLE=1 WITH_CLIENT_COMMON=ON WITH_CLIENT_INTERFACE=OFF WITH_CUPS=OFF WITH_DEBUG_ALL=OFF WITH_DEBUG_CAPABILITIES=OFF WITH_DEBUG_CERTIFICATE=OFF WITH_DEBUG_CHANNELS=OFF WITH_DEBUG_CLIPRDR=OFF WITH_DEBUG_DVC=OFF WITH_DEBUG_KBD=OFF WITH_DEBUG_LICENSE=OFF WITH_DEBUG_MUTEX=OFF WITH_DEBUG_NEGO=OFF WITH_DEBUG_NLA=OFF WITH_DEBUG_NTLM=OFF WITH_DEBUG_RAIL=OFF WITH_DEBUG_RDP=OFF WITH_DEBUG_RDPDR=OFF WITH_DEBUG_RDPEI=OFF WITH_DEBUG_RDPGFX=OFF WITH_DEBUG_REDIR=OFF WITH_DEBUG_RFX=OFF WITH_DEBUG_RINGBUFFER=OFF WITH_DEBUG_SCARD=OFF WITH_DEBUG_SND=OFF WITH_DEBUG_SVC=OFF WITH_DEBUG_SYMBOLS=OFF WITH_DEBUG_THREADS=OFF WITH_DEBUG_TIMEZONE=OFF WITH_DEBUG_TRANSPORT=OFF WITH_DEBUG_TSG=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_WND=OFF WITH_DEBUG_X11=OFF WITH_DEBUG_X11_CLIPRDR=OFF WITH_DEBUG_X11_LOCAL_MOVESIZE=OFF WITH_DEBUG_XV=OFF WITH_DSP_EXPERIMENTAL=OFF WITH_DSP_FFMPEG=OFF WITH_EVENTFD_READ_WRITE=1 WITH_FAAC=OFF WITH_FAAD2=OFF WITH_FFMPEG=TRUE WITH_FFMPEG=TRUE WITH_GFX_H264=ON WITH_GPROF=OFF WITH_GSM=OFF WITH_GSSAPI=ON WITH_GSTREAMER_0_10=OFF WITH_GSTREAMER_1_0=ON WITH_ICU=OFF WITH_IPP=OFF WITH_JPEG=OFF WITH_LAME=OFF WITH_LIBRARY_VERSIONING=ON WITH_LIBSYSTEMD=OFF WITH_MACAUDIO=OFF WITH_MACAUDIO=OFF WITH_MACAUDIO_AVAILABLE=0 WITH_MANPAGES=ON WITH_MBEDTLS=OFF WITH_OPENH264=OFF WITH_OPENSLES=OFF WITH_OPENSSL=ON WITH_OSS=ON WITH_PCSC=ON WITH_PKCS11H=ON WITH_PROFILER=OFF WITH_PULSE=OFF WITH_SAMPLE=OFF WITH_SANITIZE_ADDRESS=OFF WITH_SANITIZE_ADDRESS_AVAILABLE=1 WITH_SANITIZE_MEMORY=OFF WITH_SANITIZE_MEMORY_AVAILABLE=1 WITH_SANITIZE_THREAD=OFF WITH_SANITIZE_THREAD_AVAILABLE=1 WITH_SERVER=OFF WITH_SERVER_INTERFACE=ON WITH_SMARTCARD_INSPECT=OFF WITH_SOXR=OFF WITH_SSE2=ON WITH_SWSCALE=OFF WITH_THIRD_PARTY=OFF WITH_VALGRIND_MEMCHECK=OFF WITH_VALGRIND_MEMCHECK_AVAILABLE=1 WITH_WAYLAND=OFF WITH_X11=ON WITH_X264=OFF WITH_XCURSOR=ON WITH_XEXT=ON WITH_XFIXES=ON WITH_XI=ON WITH_XINERAMA=ON WITH_XKBFILE=ON WITH_XRANDR=ON WITH_XRENDER=ON WITH_XSHM=ON WITH_XV=ON WITH_ZLIB=ON
Build type:          Release
CFLAGS:               -fPIC -Wall -Wno-unused-result -Wno-unused-but-set-variable -Wno-deprecated-declarations -fvisibility=hidden -Wimplicit-function-declaration -Wredundant-decls -DWINPR_DLL
Compiler:            GNU, 8.3.0
Target architecture: x64
```

martinl commented:
I have Windows Server 2016 installed and smartcard authentication is working when using two different smarcards (EstEID and Yubikey4) and Windows 10+MS RDP as client.

I switched to using Heimdal kerberos on Debian 10, since that is default kerberos implementation used  for samba on Debian. Also fixed freerdp cmake build config agaist heimdal on Debian (PR is pending). FreeRDP master and informatimago:cssp-smartcard forks currently segfault when smartcard is connected while rdp smartcard redirection is active or when /smartcard-logon is specified while connecting to server. Enabling/disablng NLA did not change the sefaulting behaviour. 

```
# smartcard redirect without NLA
./client/X11/xfreerdp /v:winsrv2016-1.test.local /u:test-adm /d:test.local /smartcard -sec-nla

```
Here is one debug log from master FreeRDP repo build:
[freerdp-master.log](https://github.com/FreeRDP/FreeRDP/files/3413739/freerdp-master.log)

informatimago commented:
> On 20 Jul 2019, at 17:14, martinl <notifications@github.com> wrote:
> 
> I have Windows Server 2016 installed and smartcard authentication is working when using two different smarcards (EstEID and Yubikey4) and Windows 10+MS RDP as client.
> 
> I switched to using Heimdal kerberos on Debian 10, since that is default kerberos implementation used for samba on Debian. Also fixed freerdp cmake build config agaist heimdal on Debian (PR is pending). FreeRDP master and informatimago:cssp-smartcard forks currently segfault when smartcard is connected while rdp smartcard redirection is active or when /smartcard-logon is specified while connecting to server. Enabling/disablng NLA did not change the sefaulting behaviour.
> 
> # smartcard redirect without NLA
> ./client/X11/xfreerdp /v:winsrv2016-1.test.local /u:test-adm /d:test.local /smartcard -sec-nla
> 
> Here is one debug log from master FreeRDP repo build:
> freerdp-master.log <https://github.com/FreeRDP/FreeRDP/files/3413739/freerdp-master.log>
The log seems truncated. 

Perhaps the crash prevented it to completely flush the buffer, but this would be a bug in the log system.

I don’t see any information about the crash: 
- where did it crash?
- what is the backtrace?
- do you have a core file?


-- 
__Pascal J. Bourguignon__

martinl commented:
I made some progress, rebuilt master with debug options on Debian 10 and now smartcard authentication without NLA using EstEID card is working. However certutil -scinfo inside windows rdp session displayed reader and card info but crashed after pin request prompt. Here is the build/test info:

```
cmake -DCMAKE_BUILD_TYPE=Debug -DWITH_DEBUG_ALL=ON -DWITH_PCSC=ON -DWITH-FFMPEG=ON -DWITH_GSSAPI=ON .
cmake --build .

# smartcard logon without NLA
./client/X11/xfreerdp /v:winsrv2016-1.test.local /u:test-adm@test.local /smart-sizing /smartcard /smartcard-logon -sec-nla
```

Which tools and debug options do you recommend to use for getting useful logs for debugging/troubleshooting?

Martin

informatimago commented:
> On 21 Jul 2019, at 09:18, martinl <notifications@github.com> wrote:
> 
> I made some progress, rebuilt master with debug options on Debian 10 and now smartcard authentication without NLA using EstEID card is working. However certutil -scinfo inside windows rdp session displayed reader and card info but crashed after pin request prompt. Here is the build/test info:
> 
> cmake -DCMAKE_BUILD_TYPE=Debug -DWITH_DEBUG_ALL=ON -DWITH_PCSC=ON -DWITH-FFMPEG=ON -DWITH_GSSAPI=ON .
> cmake --build .
> 
> # smartcard logon without NLA
> ./client/X11/xfreerdp /v:winsrv2016-1.test.local /u:test-adm@test.local /smart-sizing /smartcard /smartcard-logon -sec-nla
> Which tools and debug options do you recommend to use for getting useful logs for debugging/troubleshooting?
> 

Here is the compilation script I used.  Using the sanity options is nice to find a lot of problems, before the crashes occur.

#----------------------------------------------------------------------------------------------------------------------------

#!/bin/bash -eu

freerdp_deps=(

    build-essential git-core debhelper cdbs dpkg-dev
    autotools-dev cmake pkg-config xmlto libssl-dev docbook-xsl
    xsltproc libxkbfile-dev libx11-dev libwayland-dev libxrandr-dev
    libxi-dev libxrender-dev libxext-dev libxinerama-dev libxfixes-dev
    libxcursor-dev libxv-dev libxdamage-dev libxtst-dev libcups2-dev
    libpcsclite-dev libasound2-dev libpulse-dev libjpeg-dev
    libgsm1-dev libusb-1.0-0-dev libudev-dev libdbus-glib-1-dev
    uuid-dev libxml2-dev
    libavutil-dev libavcodec-dev libavresample-dev libcunit1-dev libdirectfb-dev xmlto
    doxygen libxtst-dev


    # pkcs#11 stuff:
    opencryptoki opensc
    libopencryptoki-dev libopencryptoki0
    gnupg-pkcs11-scd
    libpkcs11-helper1-dev libpkcs11-helper1
    libp11-dev libp11-kit-dev libp11-kit0 p11-kit p11-kit-modules
    pkcs11-dump pkcs11-data

    # kerberos
    krb5-config krb5-doc krb5-locales krb5-multidev
    libkrb5-3 libkrb5-dbg libkrb5-dev libkrb5support0

    krb5-admin-server krb5-auth-dialog krb5-gss-samples  krb5-kdc krb5-kdc-ldap
    krb5-otp krb5-pkinit krb5-strength krb5-sync-plugin krb5-sync-tools krb5-user

    # gssapi
    libgssapi-krb5-2
    libgss-dbg libgss-dev libgss3

    $(~/bin/distribution | ( read sys dist release ;
	                         if [ "$dist" = ubuntu ] ; then
		                         case "$release" in
		                         (18.04)
			                         packs=(
			                             opensc-pkcs11 gnupg-pkcs11-scd-proxy
			                             libengine-pkcs11-openssl1.1

			                             ykcs11 yubico-piv-tool

			                             libkdb5-9
			                             libkadm5clnt-mit11 libkadm5srv-mit11
			                             krb5-k5tls krb5-kpropd

			                             gss-ntlmssp gss-ntlmssp-dev

			                             libgstreamer-gl1.0-0 libgstreamer-plugins-base1.0-0 libgstreamer-plugins-base1.0-dev
			                             libgstreamer-plugins-good1.0-0 libgstreamer-plugins-good1.0-dev libgstreamer1.0-0
			                             libgstreamer1.0-0-dbg libgstreamer1.0-dev libgstreamermm-1.0-1 libgstreamermm-1.0-dev
			                             libgstreamermm-1.0-doc
			                         )
			                         ;;
		                         (*)
			                         packs=(
			                             libgstreamer1.0-dev libgstreamer0.10-dev
			                             libgstreamer-plugins-base1.0-dev libgstreamer-plugins-base0.10-dev
			                         )
			                         ;;
		                         esac
	                         else
		                         packs=(
		                             libgstreamer1.0-dev libgstreamer0.10-dev
		                             libgstreamer-plugins-base1.0-dev libgstreamer-plugins-base0.10-dev
		                         )
	                         fi
	                         echo ${packs[@]}
	  ) )
)


function install_deps(){
    sudo apt-get install ${freerdp_deps[@]}
}




sanitize_ld=(
        -fsanitize=leak
)

sanitize_cc=(
    -fsanitize=address
    -fsanitize=null
    -fsanitize=bounds
    -fsanitize=vla-bound
    -fsanitize=object-size

    -fsanitize=unreachable
    -fsanitize=return # C++ only

    # -fsanitize=shift
    # -fsanitize=shift-exponent
    # -fsanitize=shift-base
    # -fsanitize=integer-divide-by-zero
    # -fsanitize=signed-integer-overflow

    -fsanitize=float-divide-by-zero
    -fsanitize=float-cast-overflow
    -fsanitize=nonnull-attribute
    -fsanitize=returns-nonnull-attribute
    -fsanitize=bool
    -fsanitize=enum
    -fsanitize=vptr # C++ only


    -fsanitize-address-use-after-scope
    -fsanitize-undefined-trap-on-error
    -fstack-protector-all
    -fstack-check
)


function trim_spaces(){
    local string="$1"
    # ${a## } or ${a%% } doesn't work! (in bash 4.3).
    string="$(echo "$string" | sed -e 's/^ *//' -e 's/ *$//')"
    echo -n "${string}"
}

function trim_colons(){
    local string="$1"
    string="$(echo "$string" | sed -e 's/^:*//' -e 's/:*$//')"
    echo -n "${string}"
}

function with_gcc_8(){
    gcc_prefix="/usr/local/gcc"
    libs=(${gcc_prefix}/lib64 ${gcc_prefix}/lib /usr/local/lib64 /usr/local/lib /usr/lib /lib )
    export PATH=$(trim_colons "${gcc_prefix}/bin:${PATH}")
    export CFLAGS=$(trim_spaces "-I${gcc_prefix}/include ${CFLAGS:-}")
    export CXXFLAGS=$(trim_spaces "-I${gcc_prefix}/include ${CXXFLAGS:-}")
    export LDFLAGS=$(trim_spaces "${LDFLAGS:-} $(printf -- '-L%s '  "${libs[@]}")")
    export LD_LIBRARY_PATH=$(trim_colons      $(printf -- '%s:'    "${libs[@]}"))
    export PKG_CONFIG_PATH="${gcc_prefix}/lib/pkgconfig:/usr/local/lib/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig:/usr/lib/x86_64-linux-gnu/pkgconfig"

    gcc -dumpversion
    case "$(gcc -dumpversion)" in
    [789]*)
        export CFLAGS=$(trim_spaces "${CFLAGS:-} ${sanitize_cc[*]} -g -g3 -ggdb -O0")
        export LDFLAGS=$(trim_spaces "${LDFLAGS:-} ${sanitize_ld[*]}")
        ;;
    *)
        export CFLAGS=$(trim_spaces "${CFLAGS:-} -g -g3 -ggdb -O0")
        export LDFLAGS=$(trim_spaces "${LDFLAGS:-}")
        ;;
    esac

    for var in PATH CFLAGS CXXFLAGS LDFLAGS LD_LIBRARY_PATH PKG_CONFIG_PATH ; do
        printf '%-24s = %s\n' "${var}" "${!var}"
    done
}

function compile_and_test(){
    cmake \
	    -DWITH_SSE2=ON    \
        -DWITH_GSSAPI=ON   \
        -DWITH_PKCS11H=ON   \
        -DWITH_KERBEROS=ON   \
        -DWITH_LIBSYSTEMD=OFF \
	    \
	    -DWITH_CUNIT=ON         \
	    -DBUILD_TESTING=ON       \
	    -DCMAKE_BUILD_TYPE=Debug  \
	    \
	    -DWITH_DEBUG_NTLM=ON \
	    -DWITH_DEBUG_NEGO=ON  \
	    -DWITH_DEBUG_NLA=ON    \
	    -DWITH_DEBUG_SCARD=ON   \
	    . \
	    && make \
	    && make CTEST_OUTPUT_ON_FAILURE=1 test
}

function main(){
    with_gcc_8

    # install_deps
    compile_and_test

    printf "\nUse: \n%s\n\n" "export ASAN_OPTIONS=verbosity=1:debug=true:check_initialization_order=true:detect_stack_use_after_return=true:strict_string_checks=true"
    exit 0
}

main "$@"

#----------------------------------------------------------------------------------------------------------------------------




-- 
__Pascal J. Bourguignon__

martinl commented:
Thanks, I will test later today

martinl commented:
I did some testing with cssp-smartcard branch and used your build script.

macOS Mojave
Both EstEID and Yubikey4 worked for RDP smartcard logon without NLA. When running certutil -scinfo inside RDP session EstEID crashed certutil -scinfo process after frist pin prompt. For Yubikey4, certutil -scinfo finished successfully

Debian 10
I switched back to using MIT kerberos dev dependencies (from heimdal-dev) and after that your cmake compile script finished successfully (first got some missing dependency errors when building against heimdal-dev)

EstEID - RDP smartcard auth without NLA works, smartcard redirect inside rdp session does not work (windows smartcard service is in uknown state error when running certutil -scinfo)
Yubikey4 - smartcard auth without NLA does not work: "No valid certificates on this smartcard were found" message on windows logon screen

I also did some initial testing with NLA support commandline options (both cards use opensc-pkcs11.so) but did not get it working yet, will continue tomorrow

martinl commented:
I did some more testing on smartcard without NLA combination in informatimago:cssp-smartcard branch.

EstEID certutil -scinfo crash in rdp session happens because EstEID has ECC certificates and certutil.exe does not support smartcard ECC certificates - crash also happens using MS RDP client. All the practical EstEID functions (client cert auth, signing, encryption) work in RDP session - tested with Digidoc4, Chrome and IE

martinl commented:
I have some test results with smartcard-logon + NLA using Yubikey4 and EstEID . My test Windows/Active Directory is using certificate name mapping, with 2 certificates mapped to one user test-adm (token label test@test.local => test-adm@test.local for Yubikey and "Firstname,Lastname,idcode" => test-adm@test.local for EstEID). UPN match cannot be used in this case since neither of the certificates has UPN in altname. 

Would you be willing to implement NLA smartcard logon using eg. token labels in addition to current UPN support?

`./client/Mac/cli/MacFreeRDP.app/Contents/MacOS/MacFreeRDP /v:${RDP_SERVER} /u:${RDP_USER} /d:${RDP_DOMAIN} /smartcard /smartcard-logon /sec:nla /T /pkcs11-module:/usr/local/lib/opensc-pkcs11.so`

Yubikey4:
```
20190724T152729.283:signal.c:112:DEBUG freerdp_handle_signals Registering signal hook...
20190724T152729.285:cmdline.c:176:DEBUG CommandLineParseArgumentsA Invalid option syntax: /v:winsrv2016-1.test.local
20190724T152729.285:cmdline.c:164:DEBUG CommandLineParseArgumentsA Option sigil not escaped: /u:test-adm
20190724T152729.285:cmdline.c:1382:DEBUG freerdp_client_detect_command_line windows: 0/1 posix: -1000/0 compat: -1000/0
20190724T152729.293:cmdline.c:3308:INFO freerdp_client_load_static_channel_addin loading channelEx rdpdr
20190724T152729.294:cmdline.c:3308:INFO freerdp_client_load_static_channel_addin loading channelEx rdpsnd
20190724T152729.298:cliprdr_main.c:1322:DEBUG cliprdr_VirtualChannelEntryEx VirtualChannelEntryEx
20190724T152729.298:cmdline.c:3308:INFO freerdp_client_load_static_channel_addin loading channelEx cliprdr
20190724T152729.306:nego.c:1209:DEBUG nego_set_negotiation_enabled Enabling security layer negotiation: TRUE
20190724T152729.306:nego.c:1221:DEBUG nego_set_restricted_admin_mode_required Enabling restricted admin mode: FALSE
20190724T152729.306:nego.c:1243:DEBUG nego_enable_rdp Enabling RDP security: FALSE
20190724T152729.306:nego.c:1255:DEBUG nego_enable_tls Enabling TLS security: FALSE
20190724T152729.306:nego.c:1267:DEBUG nego_enable_nla Enabling NLA security: TRUE
20190724T152729.306:nego.c:1279:DEBUG nego_enable_ext Enabling NLA extended security: FALSE
20190724T152729.306:nego.c:188:DEBUG nego_connect state: NEGO_STATE_NLA
20190724T152729.306:nego.c:456:DEBUG nego_attempt_nla Attempting NLA security
20190724T152729.310:tcp.c:1167:DEBUG freerdp_tcp_connect connecting to peer 192.168.0.2
20190724T152729.315:nego.c:888:DEBUG nego_send_negotiation_request RequestedProtocols: 3
20190724T152729.327:nego.c:944:DEBUG nego_process_negotiation_response RDP_NEG_RSP
20190724T152729.327:nego.c:626:DEBUG nego_recv selected_protocol: 2
20190724T152729.327:nego.c:476:DEBUG nego_attempt_nla state: NEGO_STATE_FINAL
20190724T152729.327:nego.c:203:DEBUG nego_connect Negotiated NLA security
20190724T152729.327:nego.c:252:DEBUG nego_security_connect nego_security_connect with PROTOCOL_HYBRID
20190724T152737.769:smartcard_certificate.c:313:INFO find_x509_certificates_with_signing_rsa_private_key Found 1 slots
20190724T152737.770:smartcard_certificate.c:265:INFO selected_slot Processing slot id 0
20190724T152737.774:smartcard_certificate.c:290:INFO selected_token Processing token label test@test.local (serial ecdc041bfbee2c23)
20190724T152737.774:smartcard_certificate.c:339:INFO find_x509_certificates_with_signing_rsa_private_key Opened PKCS#11 session 140224781589168
20190724T152737.774:smartcard_certificate.c:123:INFO find_x509_certificates_with_signing_rsa_private_key_in_slot Found 0 private keys
20190724T152737.775:smartcardlogon.c:42:ERROR get_info_smartcard Could not get an identity from the smartcard (nil) (reader (nil))
20190724T152737.776:nla.c:290:ERROR nla_client_init_smartcard_logon Failed to retrieve UPN !
20190724T152737.776:nla.c:604:ERROR nla_client_init Could not initialize Smartcard Logon.
20190724T152737.776:transport.c:344:ERROR transport_connect_nla NLA begin failed
20190724T152737.776:freerdp.c:853:ERROR freerdp_set_last_error freerdp_set_last_error ERRCONNECT_AUTHENTICATION_FAILED [0x00020009]
20190724T152737.776:nego.c:228:DEBUG nego_connect Failed to connect with NLA security
```

EstEID:
```
20190724T152504.340:signal.c:112:DEBUG freerdp_handle_signals Registering signal hook...
20190724T152504.342:cmdline.c:176:DEBUG CommandLineParseArgumentsA Invalid option syntax: /v:winsrv2016-1.test.local
20190724T152504.342:cmdline.c:164:DEBUG CommandLineParseArgumentsA Option sigil not escaped: /u:test-adm
20190724T152504.342:cmdline.c:1382:DEBUG freerdp_client_detect_command_line windows: 0/1 posix: -1000/0 compat: -1000/0
20190724T152504.349:cmdline.c:3308:INFO freerdp_client_load_static_channel_addin loading channelEx rdpdr
20190724T152504.351:cmdline.c:3308:INFO freerdp_client_load_static_channel_addin loading channelEx rdpsnd
20190724T152504.354:cliprdr_main.c:1322:DEBUG cliprdr_VirtualChannelEntryEx VirtualChannelEntryEx
20190724T152504.354:cmdline.c:3308:INFO freerdp_client_load_static_channel_addin loading channelEx cliprdr
20190724T152504.358:nego.c:1209:DEBUG nego_set_negotiation_enabled Enabling security layer negotiation: TRUE
20190724T152504.358:nego.c:1221:DEBUG nego_set_restricted_admin_mode_required Enabling restricted admin mode: FALSE
20190724T152504.358:nego.c:1243:DEBUG nego_enable_rdp Enabling RDP security: FALSE
20190724T152504.359:nego.c:1255:DEBUG nego_enable_tls Enabling TLS security: FALSE
20190724T152504.359:nego.c:1267:DEBUG nego_enable_nla Enabling NLA security: TRUE
20190724T152504.359:nego.c:1279:DEBUG nego_enable_ext Enabling NLA extended security: FALSE
20190724T152504.359:nego.c:188:DEBUG nego_connect state: NEGO_STATE_NLA
20190724T152504.359:nego.c:456:DEBUG nego_attempt_nla Attempting NLA security
20190724T152504.363:tcp.c:1167:DEBUG freerdp_tcp_connect connecting to peer 192.168.0.2
20190724T152504.367:nego.c:888:DEBUG nego_send_negotiation_request RequestedProtocols: 3
20190724T152504.385:nego.c:944:DEBUG nego_process_negotiation_response RDP_NEG_RSP
20190724T152504.385:nego.c:626:DEBUG nego_recv selected_protocol: 2
20190724T152504.385:nego.c:476:DEBUG nego_attempt_nla state: NEGO_STATE_FINAL
20190724T152504.385:nego.c:203:DEBUG nego_connect Negotiated NLA security
20190724T152504.385:nego.c:252:DEBUG nego_security_connect nego_security_connect with PROTOCOL_HYBRID
20190724T152516.338:smartcard_certificate.c:313:INFO find_x509_certificates_with_signing_rsa_private_key Found 2 slots
20190724T152516.340:smartcard_certificate.c:265:INFO selected_slot Processing slot id 0
20190724T152516.503:smartcard_certificate.c:290:INFO selected_token Processing token label PIN1 (xxxxxxxx,xxxxxx,xxxxxxxxxx (serial xxxxxxxxx)
20190724T152516.503:smartcard_certificate.c:339:INFO find_x509_certificates_with_signing_rsa_private_key Opened PKCS#11 session 140488345897216
20190724T152516.504:smartcard_certificate.c:123:INFO find_x509_certificates_with_signing_rsa_private_key_in_slot Found 0 private keys
20190724T152516.506:smartcard_certificate.c:265:INFO selected_slot Processing slot id 1
20190724T152516.669:smartcard_certificate.c:290:INFO selected_token Processing token label PIN2 (xxxxxxxx,xxxxxx,xxxxxxxxxx (serial xxxxxxxxx)
20190724T152516.669:smartcard_certificate.c:339:INFO find_x509_certificates_with_signing_rsa_private_key Opened PKCS#11 session 140488385503744
20190724T152516.669:smartcard_certificate.c:123:INFO find_x509_certificates_with_signing_rsa_private_key_in_slot Found 0 private keys
20190724T152516.672:smartcardlogon.c:42:ERROR get_info_smartcard Could not get an identity from the smartcard (nil) (reader (nil))
20190724T152516.672:nla.c:290:ERROR nla_client_init_smartcard_logon Failed to retrieve UPN !
20190724T152516.672:nla.c:604:ERROR nla_client_init Could not initialize Smartcard Logon.
20190724T152516.672:transport.c:344:ERROR transport_connect_nla NLA begin failed
20190724T152516.672:freerdp.c:853:ERROR freerdp_set_last_error freerdp_set_last_error ERRCONNECT_AUTHENTICATION_FAILED [0x00020009]
20190724T152516.672:nego.c:228:DEBUG nego_connect Failed to connect with NLA security
```

martinl commented:
I set up a wiki page https://github.com/martinl/FreeRDP/wiki with links to my build and test scripts in my FreeRDP test/dev fork

informatimago commented:
> On 24 Jul 2019, at 15:05, martinl <notifications@github.com> wrote:
> 
> 20190724T152737.776:nla.c:290:ERROR nla_client_init_smartcard_logon Failed to retrieve UPN !

As mentionned in some previous message, the problem is that according to the smartcard format and standard used, the way to retrieve the UPN may vary.

First, the UPN is stored in a X.509 certificate.  Those certificates can be stored (or appear to be stored) differently depending on the card format, and on the plug-in used.

Next, the way the UPN is stored in the X.509 also may vary.

Currently, it’s implemented for the IAS-ECC smartcard format.

in nla.c:288

	if (get_info_smartcard(settings) != 0)
	{
		WLog_ERR(TAG, "Failed to retrieve UPN !");
		return -1;
	}

get_info_smartcard is in  libfreerdp/core/smartcardlogon.c 
it calls scquery_X509_user_identities from the scquery module.

Note that the function scquery_X509_user_identities returns the first certificate with an UPN, but it could easily be modified to return all the certificates with a UPN: you may want to try each of them in such a case?
Currently, the UPN is found as a X.509 Subject Alt Name of type OTHERNAME, with an OID of 1.3.6.1.4.1.311.20.2.3
There’s a comment in scquery.c with the OID for the KPN if you’d want to use this instead of the UPN.
In your smartcard format, the UPN may be stored otherwise. This scquery_X509_user_identities would have to match those other formats.

So again, since it’s specific to the model of smartcard and smartcard format used, I would advise to extract these things from freerdp, take these values such as the UPN as command line argument, and use specific external tools wrapped in scripts to glue everything together (cf. my scquery tool with the sckinit script for example).

-- 
__Pascal J. Bourguignon__

martinl commented:
Thank you for the clarifications

martinl commented:
I did some initial work to split the PR into multiple upstreamable PR-s. Diffs are currently split per-directory: https://github.com/martinl/FreeRDP/tree/smartcard-logon-no-nla/diff

yoonjm commented:
It looks like there's been a lot of good progress, but it suddenly halted in July 2019. Is there any work being done on the side for this PR, or is it awaiting someone to pick up from where it's left?

martinl commented:
> It looks like there's been a lot of good progress, but it suddenly halted in July 2019. Is there any work being done on the side for this PR, or is it awaiting someone to pick up from where it's left?

I think I stopped to wait for planned major code syntax standard change in upstream which would simplify the merge. It was done but then I did not have the time to work on this for a while. In general, it would be nice if someone could pick this up and finish the merge to upstream

martinl commented:
First attempt at cleanup and breaking this into upstreamable features is at https://github.com/FreeRDP/FreeRDP/pull/6408
Any suggestions for logical order and grouping of the PR-s are welocme

yoonjm commented:
I am trying to build and test the Windows client (wfreerdp.exe) with the feature. Apart from some of the platform-specific codes that need reorganization, the only problem I had with the build was the dependency on kadm5. With that said, I believe this feature is almost ready to be supported for the Windows client. Can anyone confirm my understanding?

akallabeth commented:
@yoonjm mostly yes, but someone needs to bring this pr in shape to be mergeable. @martinl already started this, but is not there yet.
